File: Autonomous Mobile Robots/Chapter18/Lesson5/Chapter18_Lesson5.cpp
// Chapter18_Lesson5.cpp
// Case Study: Agricultural / Delivery Robots
// Composite-cost Dijkstra planner for outdoor AMR case studies

#include <iostream>
#include <vector>
#include <queue>
#include <cmath>
#include <limits>
#include <algorithm>

struct Edge {
    int u, v;
    double length_m;
    double slope_rad;
    double roughness;
    double slip_risk;
    double speed_limit;
};

double edgeTime(const Edge& e, double v_cmd) {
    double v = std::max(0.2, std::min(v_cmd, e.speed_limit));
    return e.length_m / v;
}

double edgeEnergy(const Edge& e, double mass, double v_cmd) {
    const double g = 9.81;
    double v = std::max(0.2, std::min(v_cmd, e.speed_limit));
    double c_rr = 0.03;
    double rho_air = 1.2;
    double CdA = 0.45;
    double P_aux = 80.0;

    double F_roll = c_rr * mass * g * std::cos(e.slope_rad);
    double F_grade = mass * g * std::sin(e.slope_rad);
    double F_drag = 0.5 * rho_air * CdA * v * v;
    double P = std::max(0.0, (F_roll + F_grade + F_drag) * v) + P_aux;
    return P * (e.length_m / v); // Joules
}

double edgeRisk(const Edge& e) {
    return 0.45 * e.roughness + 0.55 * e.slip_risk + 0.15 * std::abs(e.slope_rad);
}

double safeSpeedFromClearance(double clearance_m, double reaction_s, double mu, double margin = 0.5) {
    double d = std::max(0.0, clearance_m - margin);
    double a = 1.0 / (2.0 * mu * 9.81);
    double b = reaction_s;
    double c = -d;
    double disc = std::max(0.0, b*b - 4.0*a*c);
    return (-b + std::sqrt(disc)) / (2.0*a);
}

struct NodeState {
    double cost;
    int node;
    bool operator>(const NodeState& other) const { return cost > other.cost; }
};

std::pair<std::vector<int>, double> dijkstraComposite(
    const std::vector<Edge>& edges,
    int n_nodes,
    int source,
    int target,
    double mass,
    double v_cmd,
    double wt, double we, double wr,
    double reaction_s, double mu, double clearance_default
) {
    std::vector<std::vector<Edge>> g(n_nodes);
    for (const auto& e : edges) {
        g[e.u].push_back(e);
    }

    std::vector<double> dist(n_nodes, std::numeric_limits<double>::infinity());
    std::vector<int> parent(n_nodes, -1);
    dist[source] = 0.0;

    std::priority_queue<NodeState, std::vector<NodeState>, std::greater<NodeState>> pq;
    pq.push({0.0, source});

    while (!pq.empty()) {
        NodeState cur = pq.top();
        pq.pop();
        if (cur.cost > dist[cur.node]) continue;
        if (cur.node == target) break;

        for (const auto& e : g[cur.node]) {
            double v_safe = safeSpeedFromClearance(clearance_default, reaction_s, mu);
            double v_use = std::min(v_cmd, std::min(e.speed_limit, std::max(0.25, v_safe)));

            double c = wt * edgeTime(e, v_use)
                     + we * (edgeEnergy(e, mass, v_use) / 1000.0)
                     + wr * edgeRisk(e);

            double nd = dist[cur.node] + c;
            if (nd < dist[e.v]) {
                dist[e.v] = nd;
                parent[e.v] = cur.node;
                pq.push({nd, e.v});
            }
        }
    }

    std::vector<int> path;
    if (!std::isfinite(dist[target])) return {path, dist[target]};
    for (int v = target; v != -1; v = parent[v]) path.push_back(v);
    std::reverse(path.begin(), path.end());
    return {path, dist[target]};
}

int main() {
    std::vector<Edge> edges = {
        {0,1,25, 0.02,0.10,0.15,2.5},
        {1,2,18, 0.05,0.30,0.35,1.8},
        {0,3,20,-0.01,0.08,0.10,2.0},
        {3,2,24, 0.03,0.12,0.20,2.2},
        {2,4,30, 0.01,0.15,0.18,3.0},
        {3,4,28, 0.07,0.40,0.45,1.5}
    };

    auto ag = dijkstraComposite(edges, 5, 0, 4, 180.0, 1.7, 1.0, 0.03, 8.0, 0.4, 0.55, 2.5);
    std::cout << "Agricultural path: ";
    for (int n : ag.first) std::cout << n << " ";
    std::cout << " | cost = " << ag.second << "\n";

    auto del = dijkstraComposite(edges, 5, 0, 4, 45.0, 2.2, 1.5, 0.01, 5.0, 0.6, 0.65, 3.5);
    std::cout << "Delivery path: ";
    for (int n : del.first) std::cout << n << " ";
    std::cout << " | cost = " << del.second << "\n";

    double vmax = safeSpeedFromClearance(4.0, 0.5, 0.6, 0.6);
    std::cout << "Safe speed from clearance = " << vmax << " m/s\n";

    return 0;
}

File: Autonomous_Mobile_Robots_Course/Chapter18/Lesson5/Chapter18_Lesson5.py
# Chapter18_Lesson5.py
# Case Study: Agricultural / Delivery Robots
# Risk-aware mission scoring with energy, time, and traversability terms

from dataclasses import dataclass
from typing import List, Tuple
import math
import heapq


@dataclass
class Edge:
    u: int
    v: int
    length_m: float
    slope_rad: float
    roughness: float        # 0..1
    slip_risk: float        # 0..1
    speed_limit: float      # m/s


def edge_time(edge: Edge, v_cmd: float) -> float:
    v = max(0.2, min(v_cmd, edge.speed_limit))
    return edge.length_m / v


def edge_energy(edge: Edge, mass: float, v_cmd: float, c_rr: float = 0.03,
                rho_air: float = 1.2, CdA: float = 0.45, P_aux: float = 80.0) -> float:
    v = max(0.2, min(v_cmd, edge.speed_limit))
    F_roll = c_rr * mass * 9.81 * math.cos(edge.slope_rad)
    F_grade = mass * 9.81 * math.sin(edge.slope_rad)
    F_drag = 0.5 * rho_air * CdA * v * v
    P = max(0.0, (F_roll + F_grade + F_drag) * v) + P_aux
    return P * (edge.length_m / v)  # Joules


def edge_risk(edge: Edge) -> float:
    # Simple normalized traversability risk score
    return 0.45 * edge.roughness + 0.55 * edge.slip_risk + 0.15 * abs(edge.slope_rad)


def stopping_distance(v: float, reaction_s: float, mu: float, margin: float = 0.5) -> float:
    return v * reaction_s + v * v / (2.0 * mu * 9.81) + margin


def safe_speed_from_clearance(clearance_m: float, reaction_s: float, mu: float, margin: float = 0.5) -> float:
    # Solve v*t_r + v^2/(2*mu*g) + margin <= clearance
    d = max(0.0, clearance_m - margin)
    a = 1.0 / (2.0 * mu * 9.81)
    b = reaction_s
    c = -d
    disc = max(0.0, b * b - 4 * a * c)
    return (-b + math.sqrt(disc)) / (2 * a)


def build_graph(edges: List[Edge], n_nodes: int):
    g = [[] for _ in range(n_nodes)]
    for e in edges:
        g[e.u].append(e)
    return g


def shortest_path(edges: List[Edge], n_nodes: int, source: int, target: int,
                  mass: float, v_cmd: float, w_t: float, w_e: float, w_r: float,
                  reaction_s: float, mu: float, clearance_default: float = 3.0):
    g = build_graph(edges, n_nodes)
    dist = [float("inf")] * n_nodes
    parent = [-1] * n_nodes
    dist[source] = 0.0
    pq = [(0.0, source)]
    while pq:
        d, u = heapq.heappop(pq)
        if d > dist[u]:
            continue
        if u == target:
            break
        for e in g[u]:
            v_safe = safe_speed_from_clearance(clearance_default, reaction_s, mu)
            v_use = min(v_cmd, e.speed_limit, max(0.25, v_safe))
            c = (
                w_t * edge_time(e, v_use)
                + w_e * edge_energy(e, mass, v_use) / 1000.0  # kJ scaling
                + w_r * edge_risk(e)
            )
            nd = d + c
            if nd < dist[e.v]:
                dist[e.v] = nd
                parent[e.v] = u
                heapq.heappush(pq, (nd, e.v))

    # Reconstruct
    path = []
    cur = target
    if parent[cur] == -1 and cur != source:
        return [], float("inf")
    while cur != -1:
        path.append(cur)
        cur = parent[cur]
    path.reverse()
    return path, dist[target]


def agricultural_coverage_metrics(field_length: float, field_width: float,
                                  tool_width: float, overlap_ratio: float,
                                  turn_loss_s: float, cruise_speed: float) -> Tuple[int, float, float]:
    w_eff = tool_width * (1.0 - overlap_ratio)
    n_pass = math.ceil(field_width / w_eff)
    travel_dist = n_pass * field_length
    time_s = travel_dist / max(0.2, cruise_speed) + max(0, n_pass - 1) * turn_loss_s
    covered_area = field_length * field_width
    return n_pass, time_s, covered_area


def delivery_eta_chance_constraint(mean_times: List[float], std_times: List[float],
                                   deadline_s: float, z_quantile: float = 1.645) -> bool:
    mu = sum(mean_times)
    sigma = math.sqrt(sum(s * s for s in std_times))
    return (mu + z_quantile * sigma) <= deadline_s


if __name__ == "__main__":
    # Example graph shared by both robot types
    edges = [
        Edge(0, 1, 25,  0.02, 0.10, 0.15, 2.5),
        Edge(1, 2, 18,  0.05, 0.30, 0.35, 1.8),
        Edge(0, 3, 20, -0.01, 0.08, 0.10, 2.0),
        Edge(3, 2, 24,  0.03, 0.12, 0.20, 2.2),
        Edge(2, 4, 30,  0.01, 0.15, 0.18, 3.0),
        Edge(3, 4, 28,  0.07, 0.40, 0.45, 1.5),
    ]

    # Agricultural robot: heavier, risk and energy weighted
    ag_path, ag_cost = shortest_path(
        edges, n_nodes=5, source=0, target=4,
        mass=180.0, v_cmd=1.7, w_t=1.0, w_e=0.03, w_r=8.0,
        reaction_s=0.4, mu=0.55, clearance_default=2.5
    )
    print("Agricultural path:", ag_path, "cost:", round(ag_cost, 3))

    passes, t_cov, area = agricultural_coverage_metrics(
        field_length=120.0, field_width=48.0, tool_width=2.4,
        overlap_ratio=0.12, turn_loss_s=7.0, cruise_speed=1.4
    )
    print("Coverage passes:", passes, "time [min]:", round(t_cov / 60.0, 2), "area [m^2]:", area)

    # Delivery robot: lighter, time weighted, stricter deadline
    del_path, del_cost = shortest_path(
        edges, n_nodes=5, source=0, target=4,
        mass=45.0, v_cmd=2.2, w_t=1.5, w_e=0.01, w_r=5.0,
        reaction_s=0.6, mu=0.65, clearance_default=3.5
    )
    print("Delivery path:", del_path, "cost:", round(del_cost, 3))

    mean_segments = [110.0, 95.0, 150.0]
    std_segments = [12.0, 10.0, 20.0]
    ok = delivery_eta_chance_constraint(mean_segments, std_segments, deadline_s=390.0)
    print("Delivery ETA chance constraint satisfied:", ok)

    # Demonstrate safety speed bound
    vmax = safe_speed_from_clearance(clearance_m=4.0, reaction_s=0.5, mu=0.6, margin=0.6)
    print("Safe speed from clearance [m/s]:", round(vmax, 3))
